### DISCUSSION TITLE
Specifying spherical joint limits for Deep Mimic humanoid

### DISCUSSION THREAD

**Discussion by wuwingw** [bulletphysics/bullet3#3626] [General]
Hi, I'm experimenting with the Deep Mimic code and wondered if there is a way to impose limits on the spherical joints in the [humanoid.urdf](https://github.com/bulletphysics/bullet3/blob/master/examples/pybullet/gym/pybullet_data/humanoid/humanoid.urdf) in order to match those in Jason Peng's [original model](https://github.com/xbpeng/DeepMimic/blob/master/data/characters/humanoid3d.txt).

It seems that changing the .urdf file to add a `<limit>` tag (as with the revolute joints for the elbows and knees) does not make a difference, and I'm unsure if there is a way to set limits after loadURDF has been called in the code.

Thank you (and also thanks for the work in implementing Deep Mimic in pybullet, it's been very useful!)

**Comment by jhonsonlaid** [comment]
Well, I have the same problem. Could @erwincoumans help us figure this out?

**Comment by erwincoumans** [comment]
There are no limits implemented for spherical joints. Can you point to an issue that motion is out of range?  I've not seen an issue so far, since the policy matches the target motion usually well. You can enable self collision.

Let's leave the issue open in case me or someone else finds time to add them.

**Comment by erwincoumans** [comment]
It would be easiest to add the limits to this constraint btMultiBodySphericalJointMotor.cpp:
https://github.com/bulletphysics/bullet3/blob/master/src/BulletDynamics/Featherstone/btMultiBodySphericalJointMotor.cpp

**Comment by jhonsonlaid** [comment]
The agent's right hand tries to reach the target(small white box), and the shoulder rotates at a pose wihch human is hard to make. So I think I need to limit the shoulder's rotate range.
![image](https://user-images.githubusercontent.com/11907227/70680484-6bc57000-1cd3-11ea-9612-d6e9d440dbc0.png)

**Comment by arpspoof** [comment]
Hello,
Is there a schedule/plan to get this feature in the near future? I believe this is really nice to have for character animation field. A swing-twist limited spherical joint may help remove lots of unnatural movements as others have pointed out. It'll be great to see this implemented to bring improvements to character animation projects, not just deepmimic. Thanks!
